sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/partitioning/HashPartitioning.java

```diff
@@ -18,27 +18,37 @@
 package org.apache.spark.sql.connector.read.partitioning;
 
 import org.apache.spark.annotation.Evolving;
-import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.expressions.Expression;
 
 /**
- * An interface to represent data distribution requirement, which specifies how the records should
- * be distributed among the data partitions (one {@link PartitionReader} outputs data for one
- * partition).
+ * Represents a partitioning where rows are split across partitions based on the expressions
+ * returned by {@link HashPartitioning#clustering}.
  * <p>
- * Note that this interface has nothing to do with the data ordering inside one
- * partition(the output records of a single {@link PartitionReader}).
- * <p>
- * The instance of this interface is created and provided by Spark, then consumed by
- * {@link Partitioning#satisfy(Distribution)}. This means data source developers don't need to
- * implement this interface, but need to catch as more concrete implementations of this interface
- * as possible in {@link Partitioning#satisfy(Distribution)}.
- * <p>
- * Concrete implementations until now:
- * <ul>
- *   <li>{@link ClusteredDistribution}</li>
- * </ul>
+ * Data source implementations should make sure
+ * that all rows where {@link HashPartitioning#clustering} evaluate to the same value should be
+ * in the same partition.
  *
- * @since 3.0.0
+ * @since 3.3.0
  */
 @Evolving
-public interface Distribution {}
+public class HashPartitioning implements Partitioning {
+  private final Expression[] clustering;
+  private final int numPartitions;
+
```

Comment:
shall we also include the hash function (v2 function) in this class? it can be optional to represent an unknown hash algorithm.